File: algorithms/algorithms-cplusplus/leetcode/Question_0131_Palindrome_Partitioning.cc
#include <string>
#include <vector>

using namespace std;

class Solution {
 public:
  vector<vector<string>> partition(string s) {
    vector<vector<string>> res;
    vector<string> vec;
    help(s, 0, vec, res);
    return res;
  }

  void helper(string& s, size_t pos, vector<string>& vec, vector<vector<string>>& res) {
    if (pos >= s.size()) {
      res.push_back(vec);
      return;
    }
    for (size_t i = pos; i < s.size(); i++) {
      if (isPalindrome(s, pos, i)) {
        vec.push_back(s.substr(pos, i - pos + 1));
        help(s, i + 1, vec, res);
        vec.pop_back();
      }
    }
  }

  bool isPalindrome(string& s, size_t l, size_t r) {
    while (l < r) {
      if (s[l] != s[r]) return false;
      l++; r--;
    }
    return true;
  }
};

File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0131_Palindrome_Partitioning.java
package com.lanceolata.coding.algorithms.java.leetcode;

import java.util.ArrayList;
import java.util.List;

public class Question_0131_Palindrome_Partitioning {
    public List<List<String>> partition(String s) {
        List<List<String>> res = new ArrayList<>();
        if (s == null || s.length() == 0) {
            return res;
        }
        List<String> list = new ArrayList<>();
        help(s, 0, list, res);
        return res;
    }

    private void help(String s, int pos, List<String> list, List<List<String>> res) {
        if (pos >= s.length()) {
            res.add(new ArrayList<String>(list));
            return;
        }
        for (int i = pos; i < s.length(); i++) {
            if (isPalindrome(s, pos, i)) {
                String sub = s.substring(pos, i + 1);
                list.add(sub);
                help(s, i + 1, list, res);
                list.remove(list.size() - 1);
            }
        }
    }

    private boolean isPalindrome(String s, int l, int r) {
        while (l < r) {
            if (s.charAt(l++) != s.charAt(r--)) {
                return false;
            }
        }
        return true;
    }
}
